### ISSUE TITLE
protos: compile out unnecessary protos

### ISSUE THREAD

**Issue by junr03** [lyft/envoy-mobile#178]
I noticed from object file analysis that compiled protos make up the biggest percentage of binary size:

```
  27.7%  2.32Mi  27.8%  2.36Mi compiled protos
```

More importantly, there are a lot of protos that are not _needed_ in Envoy Mobile:

```
       2.4%  57.0Ki   2.4%  58.1Ki bazel-out/aarch64-opt/bin/external/envoy_api/envoy/admin/v2alpha/config_dump.pb.cc
```

However, upstream Envoy compiles all of these in, mostly because they are all needed in the server's bootstrap code. Given the size, it is worth it to investigate ways to compile out unnecessary protos. Or better yet, link in protos on-demand.

### Size Win

This entirely depends on what protos we end up needing to link in, but the ceiling is ~2mb.

**Comment by htuch** [comment]
@junr03 do you have any scripts or tooling for this kind of analysis that we could add to Envoy CI?

I'm interested in driving some automation around Envoy's dependencies, and we've added binary size to the policy in https://docs.google.com/document/d/1HbREo7pv7rgeIIjQn6mNpySzQE5rx2Yv9dXm5NqR2N8/edit#heading=h.xnpvc6pk0h0v.

It would be helpful to be able to quickly spot the binary size contribution from different libraries, and I'm guessing you might have already done some of this work.

**Comment by htuch** [comment]
Answering my own question, I think https://envoy-mobile.github.io/docs/envoy-mobile/latest/development/performance/binary_size.html captures this and we want to use Bloaty..

**Comment by junr03** [comment]
@htuch yep, that's the process I follow. Using different bloaty configs with different path regexes affords you a lot of ways to dice the binary.

Awesome! Glad to see size making it in there. Lmk if I can be of help.

**Comment by alyssawilk** [comment]
Also proto folks call out --copt='-DPROTOBUF_EXPERIMENTAL_USE_TAIL_CALL_TABLE_PARSER=1 as a new flag for improving size of files we could look into using

**Comment by alyssawilk** [comment]
also proto lite if we can avoid reflection for mobile

**Comment by fowles** [comment]
A high cost option in this space is to migrate to something based on https://github.com/protocolbuffers/upb

Protobuffer team is working on various aspects of this as part of long term strategies.  It is important to note, however, that upb is an explicitly **unsable api**.  It is intended to be an internal toolkit for building protobuf implementations.

**Comment by fowles** [comment]
gRPC already does some of this https://github.com/grpc/grpc/tree/master/third_party/upb
